### PR TITLE
syntax: Implement #![no_core]

### LINKED ISSUE #27394: Tracking issue for `no_std` (RFC #1184)
https://github.com/rust-lang/rfcs/pull/1184

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is an implementation of [RFC 1184](https://github.com/rust-lang/rfcs/pull/1184) which tweaks the behavior of
the `#![no_std]` attribute and adds a new `#![no_core]` attribute. The
`#![no_std]` attribute now injects `extern crate core` at the top of the crate
as well as the libcore prelude into all modules (in the same manner as the
standard library's prelude). The `#![no_core]` attribute disables both std and
core injection.

Closes #27394

### CODE REVIEW

File: src/libcore/prelude/v1.rs [hunk 1]
```diff
@@ -17,12 +17,6 @@
 //!
 //! There is no method to automatically inject this prelude, and this prelude is
```
Comment (kevinmehall) [inline src/libcore/prelude/v1.rs]: This sentence could be updated

File: src/test/run-pass/smallest-hello-world.rs [hunk 2]
```diff
@@ -12,10 +12,11 @@
 
 // pretty-expanded FIXME #23616
 
-#![feature(intrinsics, lang_items, start, no_std, libc)]
-#![no_std]
+#![feature(intrinsics, lang_items, start, no_core, libc, core)]
+#![no_core]
 
 extern crate libc;
+extern crate core;
```
Comment (liigo) [inline src/test/run-pass/smallest-hello-world.rs]: I don't think this is really required. Why import `liblibc` and `libcore`?

**Comment by alexcrichton** [comment]
I've also pushed a commit to make `--test` a bit more hygienic

**Comment by Gankro** [comment]
../src/libstd/sys/unix/thread.rs:169:9: 169:24 error: unused import,
#[deny(unused_imports)] on by default
../src/libstd/sys/unix/thread.rs:169     use prelude::v1::*;
                                             ^~~~~~~~~~~~~~~

**Pull request merged**